### ISSUE TITLE
Default encoding of bytes should depend on content type of request/response?

### ISSUE THREAD

**Issue by bterlson** [microsoft/typespec#2370]
Given the [following example](https://cadlplayground.z22.web.core.windows.net/?c=aW1wb3J0ICJAdHlwZXNwZWMvaHR0cCI7DQp1c2luZyBUeXBlU3BlYy5IdHRwOw0KDQpvcCB1cGxvYWRJbWFnZShAaGVhZGVyIGNvbnRlbnTEMDogIm11bHRpcGFydC9mb3JtLWRhdGEiLCBAYm9keSBpxDg6IMU%2FVcVKUmVxdWVzdCk6IHZvaWTFZm1vZGVs0yQgew0KICBAZW5jb2RlKCJiaW5hcnkiKSAvLyBzaG91bGQgdGhpcyBiZSBuZWNlc3Nhcnk%2FxDLkAII6IGJ5dGVzO8QQZmlsZU5hbWU6IHN0cmluZzsNCn0%3D&e=%40typespec%2Fopenapi3&options=%7B%7D):

```
import "@typespec/http";
using TypeSpec.Http;

op uploadImage(@header contentType: "multipart/form-data", @body image: ImageUploadRequest): void;

model ImageUploadRequest {
  @encode("binary") // should this be necessary?
  data: bytes;
  fileName: string;
}
```

Should I need to specify a `binary` encoding when the response is `multipart/form-data`? The default behavior of `format: byte` would be ridiculous for this content type.

**Comment by timotheeguerin** [comment]
We already have different behavior depending if json(base64) or octet-stream(binary) so we could expand

**Comment by mikekistler** [comment]
Each part in a multipart/form-data body has an explicit or implicit content-type. The default content type is `text/plain` ([RFC 7578](https://www.rfc-editor.org/rfc/rfc7578#section-4.4)). So if we decide that bytes is binary I think we also need to set the content type to application/octet-stream.

Your example also makes me wonder -- is the filename in ImageUploadRequest intended to a separate form field or sent in the `content-disposition` header of the field for `data`? And whichever is the answer, how would I get the other behavior?

**Comment by bterlson** [comment]
If I'm reading the OA spec correctly, it defines a default content type for each part thusly:

* If the property is a primitive, or an array of primitive values, the default Content-Type is text/plain
* If the property is complex, or an array of complex values, the default Content-Type is application/json
* If the property is a type: string with format: binary or format: base64 (aka a file object), the default Content-Type is application/octet-stream

So I think we could fix this without changing other emit?

The example is intending to represent multiple form fields (and thus, multiple parts in a multipart/form-data request). Also, content-disposition I believe is only a response header.

If we wanted to specify a request that used `content-disposition` to describe parts in the response, I think it would be sufficient to add `@header contentDisposition: string` to your response model. I suspect implementations are likely to include this header anyway regardless of what the spec says.

**Comment by markcowl** [comment]
est: 5

**Comment by lmazuel** [comment]
I'd like to understand how this is will be exposed at the TS lib level before having a strong opinion
cc @iscai-msft @m-nash

**Comment by timotheeguerin** [comment]
Getting into some issue if the same model is used as multi part and json payload

https://cadlplayground.z22.web.core.windows.net/?c=aW1wb3J0ICJAdHlwZXNwZWMvaHR0cCI7CnVzaW5nIFR5cGVTcGVjLkh0dHA7Cgptb2RlbCBBZGRyZXNzIHsKICBzdHJlZXQ6xAhpbmc7CiAgY2l0ecoQfcg1Rm9ybcUyaWTMLmHGTDrIVcQUcHJvZmlsZUltYWdlOiBieXRlc8VJQHJvdXRlKCIvbXVsdGlwYXJ0Iikgb3AgyQ8oCiAgQGhlYWRlciBjb250ZW505ADCOiAiySMvZm9ybS1kYXRhIiwKICAuLi7kAJksCik6IHZvaWQ7y2Zqc29uxmHECijWWWFwcGxpY2F0aW9uxjEsyFTJUg%3D%3D&e=@typespec/openapi3&options=%7B%7D

**Comment by qiaozha** [comment]
Running into similar question. 
I wonder if it's okay to have more than one parts that are defined as bytes and one of them could be just base64 encoded string? If yes, then we probably need to ask the author being explicit here ?

**Comment by qiaozha** [comment]
Just think of another question, to think this reversely, if it's okay to have binary bytes as model property in content type like application/json and text/plain ? see the [playground link here](https://cadlplayground.z22.web.core.windows.net/?c=aW1wb3J0ICJAdHlwZXNwZWMvaHR0cCI7Cgp1c2luZyBUeXBlU3BlYy5IdHRwOwpAc2VydmljZSh7CiAgdGl0bGU6ICJXaWRnZXQgU8YbIiwKICB2ZXJzaW9uOiAiMS4wLjAiLAp9KQpuYW1lc3BhY2UgRGVtb8cvOwoKbW9kZWwgx0bEWUB2aXNpYmlsaXR5KCJyZWFkIiwgInVwZGF0ZSIpxCBwYXRoCiAgaWQ6IHN0cmluZzsKCiAgd2VpZ2h0OiBpbnQzMjsKICBjb2xvcjogInJlZCIgfCAiYmx1ZeQA2yAgQGVuY29kZSgiYmluYXJhecVXdGVzdEJ5dGVzOiBixAc7Cn0KCkBlcnJvcucApEXEDOUAo8Q%2By2ptZXNzYWdl6gCPxDpyb3V0ZSgiL3flANpzIikKQHRhZyjnATDEEGludGVyZuQBEscUxVxAcG9zdCBjcmVhdGUoLi4uxhwpOsclIHzmAIjEZQ%3D%3D&e=%40typespec%2Fopenapi3&options=%7B%22linterRuleSet%22%3A%7B%22extends%22%3A%5B%22%40typespec%2Fhttp%2Fall%22%5D%7D%7D) 

If it's no, then the current default content type for models with binary properties might be incorrect and multipart/form-data might be more appropriate for this case ?

**Comment by timotheeguerin** [comment]
I don't think it make sense if you saying it is kept as binary data(with that `@encode` which also is not a known encoding - it has just been used to make the openapi look the same).

However using `bytes` inside of json is totally valid, they just need to be encoded as something that is a `string`. By default it is `base64`

**Comment by qiaozha** [comment]
here binary bytes, I mean more like a binary stream. if it's base64 encoded bytes, it just a normal string in the wire layer which is allowed without any doubts here.
In openapi3 documentation https://swagger.io/docs/specification/describing-responses/#response-that-returns-a-file, it does say that file could be embedded into a json or xml as a base64 encoded string, but it's not clear about whether file could embedded into a json or xml as binary stream.

**Comment by timotheeguerin** [comment]
that's what I mean, Json is a string format you cannot have some random binary bytes in it, it wouldn't parse. Everything needs to be valid json(a number, a boolean, a string, an array or an object)

So you can have files be sent over json but then need to be serialized/encoded with base64(or some other string encoding)

**Comment by qiaozha** [comment]
Got it, Thanks :)
That is kind of consistent with what I said earlier https://github.com/microsoft/typespec/issues/2370#issuecomment-1760712181 ?  if we have a bytes marked as @encode("binary") or binary bytes as you mentioned inside a model, we actually should not use application/json as the default content type ?
